flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.channel;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.state.StateObject;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+/**
+ * Writes channel state during checkpoint/savepoint.
+ */
+@Internal
+public interface ChannelStateWriter extends AutoCloseable {
+
+	/**
+	 * Initiate write of channel state for the given checkpoint id.
+	 */
+	void start(long checkpointId);
+
+	/**
+	 * Add in-flight buffers from the {@link org.apache.flink.runtime.io.network.partition.consumer.InputChannel InputChannel}.
+	 * <p>Must be called after {@link #start} and before {@link #finish(long)}.</p>
+	 * @param startSeqNum is the sequence number of the 1st passed buffer
+	 *                    It’s intended use is incremental snapshots
+	 *                    For the restored buffers it should be -1
```

Comment:
I'd don't get the idea of having -1 here. Doesn't it make usage more complicated?
I thought adding the real sequence number would make incremental checkpointing easier (you just need to remember the highest number of the previous checkpoint).

Reply:
For incremental checkpointing, we need to distinguish somehow restored buffers from new buffers. Given that:
* sequence numbers currently start from 0 after the restart
* buffers don't carry any sequence numbers themselves
we can't compare the given argument with something from the previous run.

Can you please explain what are "real sequence numbers"?

Reply:
I thought about storing the sequence numbers alongside the data. But I haven't thought about sequence numbers being reset.
I'm simply not sure how incremental checkpoints should work if we don't have that information. 

Let's assume there is a checkpoint where for one specific channel buffer 0..9 are stored. Now, there is a recovery where only 0..5 are completely unspilled before a new checkpoint is taken, in which 6..15 are stored. For incremental snapshot, we would simply skip storing 6..9 because at the current API, they'd get a -1. However, how would the recovery know where to resume?

Reply:
I assume that by unspilling you mean sending the buffers out (to the operator or to downstream); and that we have 6 new buffers.

The other one is to count how many buffers or bytes actually are in-flight and store this offset.
One option would be to do a full snapshot if there any negative sequence numbers.

Of course, having SNs inside the buffers would simplify a lot, but I guess it would require more effort and discussion.